fossas/spectrometer
# Compatibility script: `fossa report attribution --json`

# Overview

Adds a compatibility script for `fossa attribution --json`.

## Acceptance criteria

* The compatibility layer transforms the report to use the same key structure and content.

The sole exception to this is that FOSSAv1 sometimes outputs `Notes` keys (which are arrays of strings) with a single zero-value string, but sometimes not. The FOSSAv2 version of the report consistently reports an empty array in these cases, so we don't really have a way to reproduce this exactly.

Example:

<img width="1048" alt="Screen Shot 2021-10-18 at 5 09 40 PM" src="https://user-images.githubusercontent.com/55713231/137822920-2c469384-47c9-4263-9897-89389ef24d62.png">

## Testing plan

* Run a scan on a project.
* Run `fossa report attribution --json` on that same project using both FOSSA v1 and v2. Save their outputs to disk and note the differences.
* Run `fossa report attribution --json | go build scripts/compat-attribution/main.go` using FOSSA v2. Save its output to disk.
* Note that the difference between the compat version of the output and the v1 output is within AC.

## Risks

Not very risky.

## Potentially Contentious Decisions

I set up a Go project at the root of Spectrometer to make this happen.

I originally spent a fair amount of time trying to hack this with something like `jq` in pure bash because I wanted to avoid this, but it quickly became a mess. Rather than go down that route I decided that this was the more reasonable option. Happy to discuss it with the team if there are strong objections though.

This is also why I moved the `vendor/` folder to `vendor-bins/`: the existence of a `vendor/` folder made Go very sad indeed.

## Checklist


Also I didn't include any tests because:

* The operation of this script is _extremely_ straightforward and mechanical.
* I don't see a good way for us to integration test against an actual FOSSA instance, but if I'm missing something I'd love to be shown how I can do it.
